src/ciadmin/generate/tcyml.py

```diff
@@ -27,53 +29,85 @@ async def get(repo_path, repo_type="hg", revision=None, default_branch=None):
         if revision is None:
             revision = default_branch or "default"
         url = f"{repo_path}/raw-file/{revision}/.taskcluster.yml"
+        cache_key = (url, revision)
+
+        async with _lock.setdefault(cache_key, Lock()):
+            if cache_key in _cache:
+                return _cache[cache_key]
+
+            client = RetryClient(
+                client_session=aiohttp_session(),
+                # Despite only setting 404 here, 5xx statuses will still be retried
+                # for. See https://github.com/inyutin/aiohttp_retry?tab=readme-ov-file
+                # for details.
+                retry_options=ExponentialRetry(attempts=5, statuses={404}),
+            )
+            headers = {"User-Agent": USER_AGENT}
+            params = {}
+            async with client.get(url, headers=headers, params=params) as response:
+                try:
+                    response.raise_for_status()
+                    result = await response.read()
+                except aiohttp.ClientResponseError as e:
+                    print(f"Got error when querying {url}: {e}")
+                    raise e
+
+            _cache[cache_key] = result
+
     elif repo_type == "git":
         if revision is None:
             revision = default_branch or "master"
         if repo_path.startswith("https://github.com/"):
             if repo_path.endswith("/"):
                 repo_path = repo_path[:-1]
             repo = repo_path.replace("https://github.com/", "")
-            url = f"https://api.github.com/repos/{repo}/contents/.taskcluster.yml"
+            endpoint = f"/repos/{repo}/contents/.taskcluster.yml"
         elif repo_path.startswith("git@github.com:"):
             if repo_path.endswith(".git"):
                 repo_path = repo_path[:-4]
             repo = repo_path.replace("git@github.com:", "")
-            url = f"https://api.github.com/repos/{repo}/contents/.taskcluster.yml"
+            endpoint = f"/repos/{repo}/contents/.taskcluster.yml"
         else:
             raise Exception(
                 f"Don't know how to determine file URL for non-github repo: {repo_path}"
             )
+
+        cache_key = (endpoint, revision)
+
+        async with _lock.setdefault(cache_key, Lock()):
+            if cache_key in _cache:
+                return _cache[cache_key]
+
+            if "GITHUB_TOKEN" in os.environ:
+                client_cls = partial(TokenClient, os.environ["GITHUB_TOKEN"])
+            elif "GITHUB_APP_ID" in os.environ and "GITHUB_APP_PRIVKEY" in os.environ:
+                client_cls = partial(
+                    AppClient,
+                    os.environ["GITHUB_APP_ID"],
+                    os.environ["GITHUB_APP_PRIVKEY"],
+                    owner="mozilla-releng",
+                    repositories=["fxci-config"],
+                )
+            else:
+                client_cls = PublicClient
```

Comment:
It might be nice to add a `simple_github.from_env` helper function that does this